### PR TITLE
Integration tests staking v4

### PR DESCRIPTION
**Pull request by mariusmihaic**
What's new
--
- `StakingV4DistributeAuctionToWaitingEpoch` activation flag = epoch in which selected nodes from auction will be distributed to waiting list (_used in node shuffler_)
- Added a new entry in `MaxNodesChangeEnableEpoch` config, which will have enable epoch = `StakingV4DistributeAuctionToWaitingEpoch` and will decrease the max nodes config with the number of shuffled out validators from all shards (e.g. from 3200 -> 2880, since we shuffle out 80 eligible validators per 4 shards)
- Computation of available slots for auction nodes which will be selected based on topup as following: 
```
availableSlots = maxNodes - currentNumberOfValidators - numOfNodesWhichWillBeShuffledOut
```

Integration tests
--
- Added `TestMetaProcessor` struct which will create a predefined nodes configuration with provided params: 	
```
        numOfMetaNodes uint32,
	numOfShards uint32,
	numOfEligibleNodesPerShard uint32,
	numOfWaitingNodesPerShard uint32,
	numOfNodesToShufflePerShard uint32,
	shardConsensusGroupSize int,
	metaConsensusGroupSize int,
	numOfNodesInStakingQueue uint32,
```
1. All generated address have a fixed length which starts from index = 0 -> maxNumNodes (e.g. `00address-0` -> `0address-50`)
2. Every node in waiting/eligible is associated with a different owner
3. All nodes in staking queue are owned by a single owner
4. Every node has the same top-up per node
5. It can `Process` **n** number of rounds by committing blocks in each of them and will display nodes config each time.

- Added 2 integration tests:
1. `TestStakingV4` which simulates a mainnet nodes configuration (400 meta nodes + 400 eligible and waiting/shard + 60 nodes in staking queue, each of them having the same topup per node) going through stakingV4Init, stakingV4Enable, stakingV4DistributeAuctionToWaiting epochs.
2. `TestStakingV4MetaProcessor_ProcessMultipleNodesWithSameSetupExpectSameRootHash` which simulates multiple `TestMetaProcessor` starting to process **n** epochs with the same initial setup. It is expected that there is not root hash mismatch at any epoch

Bugfixes
---
- Deactivated `flagCorrectNumNodesToStake` flag in `legacySystemSCs.go` upon staking v4 enable epoch, since it used code related to staking queue (which is now obsolete) 
- Fixed a bug in which two nodes with the same top up would not be correctly sorted/compared by `pubKey XOR randomness`, resulting in root hash mismatches between meta shard nodes
- Uint32 substraction underflow when computing available slots from auction list
- `setNodesPerShards` in `indexHashedNodesCoordinatorLite.go` to also save shuffled out nodes
- Race condition in already existing code from `process/block/displayMetaBlock.go` by adding a mutex for `peakTPS`

Refactor
--
- Moved some testing code related to staking from `systemSCs_test.go` to `testscommon/stakingCommon.go` to be used in stakingV4 integration tests
- `NewFeeAccumulator` constructor to return directly a new instance of `*feeHandler` instead of `(*feeHandler, error)` , since it could have never returned an error
- Several stubs so that code in tests would not panic because of uninitialized memory

### CODE REVIEW

File: integrationTests/vm/staking/componentsHolderCreator.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/common"
+	"github.com/ElrondNetwork/elrond-go/common/forking"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever/blockchain"
+	"github.com/ElrondNetwork/elrond-go/epochStart/notifier"
+	factory2 "github.com/ElrondNetwork/elrond-go/factory"
+	"github.com/ElrondNetwork/elrond-go/integrationTests"
+	mock2 "github.com/ElrondNetwork/elrond-go/integrationTests/mock"
+	factory3 "github.com/ElrondNetwork/elrond-go/node/mock/factory"
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply integrationTests/vm/staking/componentsHolderCreator.go]: Renamed them

File: integrationTests/vm/staking/metaBlockProcessorCreator.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/data"
+	"github.com/ElrondNetwork/elrond-go-core/data/block"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/epochStart"
+	"github.com/ElrondNetwork/elrond-go/epochStart/metachain"
+	factory2 "github.com/ElrondNetwork/elrond-go/factory"
+	mock2 "github.com/ElrondNetwork/elrond-go/integrationTests/mock"
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply integrationTests/vm/staking/metaBlockProcessorCreator.go]: Renamed them

File: integrationTests/vm/staking/nodesCoordiantorCreator.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+import (
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/common"
+	"github.com/ElrondNetwork/elrond-go/config"
+	factory2 "github.com/ElrondNetwork/elrond-go/factory"
+	mock2 "github.com/ElrondNetwork/elrond-go/integrationTests/mock"
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply integrationTests/vm/staking/nodesCoordiantorCreator.go]: Renamed them

File: integrationTests/vm/staking/systemSCCreator.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+import (
+	"bytes"
+	"strconv"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/epochStart/metachain"
+	mock3 "github.com/ElrondNetwork/elrond-go/epochStart/mock"
+	factory2 "github.com/ElrondNetwork/elrond-go/factory"
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply integrationTests/vm/staking/systemSCCreator.go]: Renamed them

File: integrationTests/vm/staking/testMetaProcessor.go [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/data"
+	"github.com/ElrondNetwork/elrond-go-core/data/block"
+	"github.com/ElrondNetwork/elrond-go-core/display"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/common"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/epochStart/metachain"
+	factory2 "github.com/ElrondNetwork/elrond-go/factory"
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply integrationTests/vm/staking/testMetaProcessor.go]: Renamed them

File: integrationTests/vm/staking/stakingV4_test.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+	}
+}
+
+func requireMapDoesNotContain(t *testing.T, m map[uint32][][]byte, s [][]byte) {
+	for _, elemInSlice := range s {
+		require.False(t, searchInMap(m, elemInSlice))
+	}
+}
+
+// TODO: Staking v4: more tests to check exactly which nodes have been selected/unselected from previous nodes config auction
```
Comment (AdoAdoAdo) [inline integrationTests/vm/staking/stakingV4_test.go:68]: also try with different configurations of previous nodes config (before activation)

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: testscommon/rewardsCreatorStub.go [hunk 7]
```diff
@@ -65,7 +66,7 @@ func (rcs *RewardsCreatorStub) GetLocalTxCache() epochStart.TransactionCacher {
 	if rcs.GetLocalTxCacheCalled != nil {
 		return rcs.GetLocalTxCacheCalled()
 	}
-	return nil
+	return dataPool.NewCurrentBlockPool()
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply testscommon/rewardsCreatorStub.go:69]: Changed it as suggested

File: testscommon/stakingcommon/stakingCommon.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+	accountsDB state.AccountsAdapter,
+	ownerAddress []byte,
+	rewardAddress []byte,
+	stakedKeys [][]byte,
+	totalStake *big.Int,
+	marshaller marshal.Marshalizer,
+) {
+	AddValidatorData(accountsDB, ownerAddress, stakedKeys, totalStake, marshaller)
+	AddStakingData(accountsDB, ownerAddress, rewardAddress, stakedKeys, marshaller)
+	_, _ = accountsDB.Commit()
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply testscommon/stakingcommon/stakingCommon.go:28]: Added logger + logged error

File: factory/coreComponents.go [hunk 9]
```diff
@@ -310,14 +310,16 @@ func (ccf *coreComponentsFactory) Create() (*coreComponents, error) {
...
-		ShuffleBetweenShards:           true,
-		MaxNodesEnableConfig:           ccf.epochConfig.EnableEpochs.MaxNodesChangeEnableEpoch,
-		BalanceWaitingListsEnableEpoch: ccf.epochConfig.EnableEpochs.BalanceWaitingListsEnableEpoch,
-		WaitingListFixEnableEpoch:      ccf.epochConfig.EnableEpochs.WaitingListFixEnableEpoch,
+		NodesShard:                               genesisNodesConfig.MinNumberOfShardNodes(),
+		NodesMeta:                                genesisNodesConfig.MinNumberOfMetaNodes(),
+		Hysteresis:                               genesisNodesConfig.GetHysteresis(),
+		Adaptivity:                               genesisNodesConfig.GetAdaptivity(),
+		ShuffleBetweenShards:                     true,
+		MaxNodesEnableConfig:                     ccf.epochConfig.EnableEpochs.MaxNodesChangeEnableEpoch,
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply factory/coreComponents.go:318]: Yes, refactored code to use `EnableEpochs` struct in this struct

File: epochStart/metachain/systemSCs.go [hunk 10]
```diff
@@ -236,14 +268,34 @@ func (s *systemSCProcessor) getValidatorTopUpMap(validators []state.ValidatorInf
 }
 
 func compareByXORWithRandomness(pubKey1, pubKey2, randomness []byte) bool {
-	minLen := core.MinInt(len(pubKey1), len(randomness))
+	lenPubKey := len(pubKey1)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply epochStart/metachain/systemSCs.go:271]: Right, created a new func `calcNormRand` which handles randomness normalization and passed the result in `compareByXORWithRandomness`

File: epochStart/metachain/systemSCs.go [hunk 11]
```diff
@@ -236,14 +268,34 @@ func (s *systemSCProcessor) getValidatorTopUpMap(validators []state.ValidatorInf
...
+
+	rnd := randomness
+	pk1 := pubKey1
+	pk2 := pubKey2
+
+	if lenPubKey > lenRand {
+		rnd = bytes.Repeat(randomness, repeatedCt)
+		rnd = rnd[:maxLen]
+	} else {
+		pk1 = bytes.Repeat(pk1, repeatedCt)
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply epochStart/metachain/systemSCs.go:286]: Right, very good point !
Applied your suggestion, pub keys are not touched anymore, only the randomness

File: integrationTests/vm/staking/stakingV4_test.go (hunk 6 above)
Comment (mariusmihaic) [reply integrationTests/vm/staking/stakingV4_test.go:68]: Right, a different custom setup should be done here

File: epochStart/metachain/legacySystemSCs.go [hunk 12]
```diff
@@ -615,10 +619,6 @@ func (s *legacySystemSCProcessor) updateMaxNodes(validatorsInfoMap state.ShardVa
 		return err
 	}
 
-	if maxNumberOfNodes < prevMaxNumberOfNodes {
-		return epochStart.ErrInvalidMaxNumberOfNodes
-	}
-
 	if s.flagStakingQueueEnabled.IsSet() {
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply epochStart/metachain/legacySystemSCs.go:622]: I would prefer to not touch the flagStakingQueueEnabled with setting logic that is not obvious in the configuration toml.

File: epochStart/metachain/systemSCs.go [hunk 13]
```diff
@@ -267,35 +273,32 @@ func (s *systemSCProcessor) getValidatorTopUpMap(validators []state.ValidatorInf
...
 
-func compareByXORWithRandomness(pubKey1, pubKey2, randomness []byte) bool {
-	lenPubKey := len(pubKey1)
+func calcNormRand(randomness []byte, expectedLen int) []byte {
 	lenRand := len(randomness)
-
-	minLen := core.MinInt(lenPubKey, lenRand)
-	maxLen := core.MaxInt(lenPubKey, lenRand)
-	repeatedCt := maxLen/minLen + 1
+	minLen := core.MinInt(expectedLen, lenRand)
```
Comment (AdoAdoAdo) [inline epochStart/metachain/systemSCs.go:278]: why do you need minLen and maxLen?
you just need randLen and expectedLen

also can you rename lenRand to randLen?

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: epochStart/metachain/systemSCs.go (hunk 13 above)
Comment (mariusmihaic) [reply epochStart/metachain/systemSCs.go]: Right, this code could be simplified. Changed it

**Review by AdoAdoAdo — APPROVED** [review]

File: epochStart/metachain/systemSCs.go [hunk 14]
```diff
@@ -179,6 +203,14 @@ func (s *systemSCProcessor) selectNodesFromAuctionList(validatorsInfoMap state.S
 	return nil
 }
 
+// TODO: Move this in elrond-go-core
+func safeSub(a, b uint32) (uint32, error) {
```
Comment (sasurobert) [inline epochStart/metachain/systemSCs.go:207]: we have this for Uint64

**Review by sasurobert — APPROVED** [review]

File: epochStart/metachain/systemSCs.go (hunk 14 above)
Comment (mariusmihaic) [reply epochStart/metachain/systemSCs.go:207]: Right, but I didn't want to cast to uint64, but have a separate function for it. Will move it in elrond-go-core in a separate task

**Pull request merged**